### PR TITLE
LPD-42068 Decoupling error handling infrastructure from JaxRS for more consistency in batch error messages

### PR DESCRIPTION
**Pull request by albertojml**
**What's new?:**
- New problem handling infrastructure has been created to decouple the error-retrieving mechanism from JaxRS.
- Batch engine error mechanisms have been adapted to this new behavior to retrieve the expected messages.
- New tests have been added in order to check the proper behaviour of the new functionality.

This PR has expanded what was purposed in this [previous proof of concept](https://github.com/liferay-headless/liferay-portal/pull/2315). See the following **Jira Spike Ticket**: [LPD-39980](https://liferay.atlassian.net/browse/LPD-39980)

See the following **Jira Ticket** for more details: [LPD-41526](https://liferay.atlassian.net/browse/LPD-41527)

### CODE REVIEW

**Comment by albertojml** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import com.liferay.portal.vulcan.problem.ProblemMapper;
+
+import java.util.List;
+import java.util.Locale;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Luis Miguel Barcos
```
[the remark this answers is not in the document]
Comment (4lejandrito) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java:25]: :cry:

**Comment by ccorreagg** [comment]
@albertojml  There is a conflict

**Comment by 4lejandrito** [comment]
I'll fix the conflict.

**Comment by 4lejandrito** [comment]
Conflicts fixed.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+					"objectFieldName",
+					objectValidationRuleResult.getObjectFieldName()
+				));
+		}
+
+		return new Problem() {
+
+			@Override
+			public String getDetail(Locale locale) {
+				return jsonArray.toString();
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java:87]: I know this was developed in the ExceptionMapper, but it is weird to create a JSON when the response could be a XML

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+	public Status getStatus(Locale locale);
+
+	public String getTitle(Locale locale);
+
+	public String getType();
+
+	public static enum Status {
+
+		BAD_REQUEST(400, "Bad Request");
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:25]: Why are we inserting the 400 http code here? If we want to decouple, maybe we should get rid of it, at least in here

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Locale;
+
+/**
+ * @author Alejandro Tardín
+ */
+public interface Problem {
+
+	public String getDetail(Locale locale);
+
+	public Status getStatus(Locale locale);
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:17]: Does the status depends on the locale? :thinking:

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.List;
+import java.util.Locale;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Luis Miguel Barcos
+ */
+@Component(service = ProblemMapper.class)
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java:27]: With this, we are registering a component that will be used for all the REST APIs. I guess you are aware of it, right?

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemMapper.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.portal.vulcan.problem;
+
+/**
+ * @author Alejandro Tardín
+ */
+public interface ProblemMapper<T extends Throwable> {
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemMapper.java:11]: Why are you including in the ProblemProvider the @ProviderType annotation and not here?

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/ProblemProviderImpl.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * @author Alejandro Tardín
+ */
+@Component(service = ProblemProvider.class)
+public class ProblemProviderImpl implements ProblemProvider {
+
+	@Override
+	public Problem getProblem(Throwable throwable) {
+		while (throwable != null) {
+			Class<? extends Throwable> clazz = throwable.getClass();
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/ProblemProviderImpl.java:30]: Mmm, what if there is a ProblemProvider of a superclass of the exception? I think we should not be considering it, right?

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemProvider.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+package com.liferay.portal.vulcan.problem;
+
+import org.osgi.annotation.versioning.ProviderType;
+
+/**
+ * @author Alejandro Tardín
+ */
+@ProviderType
+public interface ProblemProvider {
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemProvider.java:14]: This class is providing us Problems :skull:

**Review by ccorreagg — COMMENTED** [review]
Please have a look at my comments. It is just a quick review, I will keep reviewing

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemMapper.java (hunk 6 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemMapper.java:11]: The @ProviderType goes into the framework interface to allow us to make breaking changes. This interface is meant to be implemented by external modules so it should follow the standard semver rules.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java (hunk 2 above)
Comment (4lejandrito) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java:87]: We are keeping the same behaviour, just refactoring the code. But you are right it is weird. Any ideas on how to make it better?

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 3 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:25]: This is what I wanted to review with you. I agree.

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 4 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:17]: There is a reason phrase that could be localized.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java (hunk 5 above)
Comment (4lejandrito) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java:27]: Yes, we already discussed this. Do you think we should improve this now and add some kind of filtering, or do you think we could do that later?

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/ProblemProviderImpl.java (hunk 7 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/ProblemProviderImpl.java:30]: Hmm I think we should also consider super classes... but for now this is enough for this use case. We can discuss it.

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 3 above)
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:25]: I would say, let's get rid of this in here and move it to an independant class inside the jaxrs package

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 4 above)
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:17]: Yes, but the status?

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemMapper.java (hunk 6 above)
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/ProblemMapper.java:11]: You mean external customers?

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/ProblemProviderImpl.java (hunk 7 above)
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/ProblemProviderImpl.java:30]: ok, let's just have it in mind for the future

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+public class ObjectValidationRuleEngineProblemMapper
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java:28]: I think this should be renamed from `ObjectValidationRuleEngineProblemMapper` to `ObjectValidationRuleEngineExceptionProblemMapper`

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 3 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:25]: Ok I'll simplify this as discussed.

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 4 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:17]: As discussed we will remove it.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java (hunk 9 above)
Comment (4lejandrito) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java]: Done.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java (hunk 2 above)
Comment (4lejandrito) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/vulcan/problem/ObjectValidationRuleEngineProblemMapper.java]: We will leave it like this for now, as discussed on slack.

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 3 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java]: Done!

File: modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new AggregateTestRule(
+			new LiferayIntegrationTestRule(),
+			PermissionCheckerMethodTestRule.INSTANCE);
+
+	@Test
+	public void testPostImportTask() throws Exception {
+		try (LogCapture logCapture = LoggerTestUtil.configureLog4JLogger(
```
Comment (ccorreagg) [inline modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java:47]: Should we move this to the `ExportImportTaskUtil.postImportTask` method when the expected status is FAILED?

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java:17]: I think this is not solved yet

File: modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+				).build());
+
+			Assert.assertEquals(3, (int)importTask.getProcessedItemsCount());
+
+			FailedItem[] failedItems = importTask.getFailedItems();
+
+			Assert.assertEquals(
+				Arrays.toString(failedItems), 3, failedItems.length);
+
+			for (FailedItem failedItem : failedItems) {
```
Comment (ccorreagg) [inline modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java:110]: I am thinking about the missing assertion of the HTTP code :thinking: but maybe we could include it in the Portal Vulcan TESTS that are still pending in this PR

File: modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/util/ExportImportTaskUtil.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+		while (true) {
+			exportTask = ExportTaskSerDes.toDTO(
+				HTTPTestUtil.invokeToString(
+					null,
+					"headless-batch-engine/v1.0/export-task/by-external-" +
+						"reference-code/" + externalReferenceCode,
+					Http.Method.GET));
+
+			if (!StringUtil.equals(
+					exportTask.getExecuteStatusAsString(), "STARTED")) {
```
Comment (ccorreagg) [inline modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/util/ExportImportTaskUtil.java:51]: I think the INITIAL status is also possible. Should we replace with COMPLETED / FAILED instead?

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java (hunk 11 above)
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/problem/Problem.java]: Done! Sorry.

File: modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java (hunk 12 above)
Comment (4lejandrito) [reply modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java:110]: Feel free to add any assertions here too.

File: modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/util/ExportImportTaskUtil.java (hunk 13 above)
Comment (4lejandrito) [reply modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/util/ExportImportTaskUtil.java:51]: Hmm... but that would only work if the expectedExecuteStatus is COMPLETED or FAILED. That may not always be the case.

Comment (ccorreagg) [reply modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/util/ExportImportTaskUtil.java]: but are we expecting to have a different expected status from COMPLETED or FAILED? there is only one status more, INITIAL, and there are no tests now that are expecting that status

File: modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java (hunk 10 above)
Comment (ccorreagg) [reply modules/apps/headless/headless-batch-engine/headless-batch-engine-test/src/testIntegration/java/com/liferay/headless/batch/engine/resource/v1_0/test/ImportTaskResourceTest.java:47]: I checked and there are two different logCaptures, based on the on fail strategy. I am leaving this for now

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
I have added two extra commits ([this one](https://github.com/liferay-headless/liferay-portal/pull/2347/commits/459173eea4fc821881bf817a2820740d41803c1e) and [this one](https://github.com/liferay-headless/liferay-portal/pull/2347/commits/15ccfca7e930bcc69c0c4ac980e0d8d4bbdd3d7a)). Please review @4lejandrito

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:forward

**Comment by ccorreagg** [comment]
ci:forward

**Comment by ccorreagg** [comment]
The errors do not seem related

**Comment by ccorreagg** [comment]
ci:forward:force

**Pull request closed without merging**